### PR TITLE
Inversion option

### PR DESCRIPTION
**Pull request by jmorganmartin**
My editor is set to automatically remove trailing whitespace, so a lot of lines of changes are just that, unfortunately (I love the "hide whitespace" option and wish it was the default).

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ const triadQualities = {
 };
 const legalClefs = ['treble', 'bass', 'alto', 'tenor'];
 const legalTriads = ['M', 'm', 'aug', 'dim', 'sus2', 'sus4'];
+const legalInversions = ['inv0', 'inv1', 'inv2'];
```
Comment (jmorganmartin) [inline src/App.js:21]: The param names are mostly-arbitrary. In this case, I did use string replace to fetch the number at the end, but there are many ways to achieve the same thing.

File: src/App.js [hunk 2]
```diff
@@ -65,27 +66,48 @@ const util = {
...
     return arr[util.getRand(0, arr.length - 1)];
   },
-  
+
+  rotateElements(arr, k) {
+    const n = k % arr.length;
+    if (n === 0) return arr;
+    arr.unshift( ...arr.splice(-n));
+    return arr;
+  },
```
Comment (jmorganmartin) [inline src/App.js:90]: Helper function for inverting chords. 

There is an "accidental bug"/quirk in Teoria: if the notes in the "voicing" array aren't in the correct order, accidentals are applied to the incorrect notes when rendered visually (although all other parts of the Chord class appear to work just fine with the voicing out of order), details here: https://github.com/0xfe/vexflow/issues/104  

I found this algorithm on the internet. It's not the simplest to follow, but it works.

File: src/App.js [hunk 3]
```diff
@@ -65,27 +66,48 @@ const util = {
...
+    }
+    if (j === 2 && k !== 12) {
+      return num + "nd";
+    }
+    if (j === 3 && k !== 13) {
+      return num + "rd";
+    }
+    return num + "th";
+  },
+
```
Comment (jmorganmartin) [inline src/App.js:105]: Used by the flashcard to spell the inversion correctly (1st, 2nd, 3rd, etc)

File: src/App.js [hunk 4]
```diff
@@ -158,10 +188,45 @@ function updateSettings(e) {
...
 
 function getCount() {
   return null;
 }
+
+function getInversion(chord, inv) {
+
+  const voices = chord.simple().length;
+  inv = inv % voices;
+  if (inv === 0) return chord;
```
Comment (jmorganmartin) [inline src/App.js:206]: Sanity check, root inversion = no inversion.

File: src/App.js [hunk 5]
```diff
@@ -158,10 +188,45 @@ function updateSettings(e) {
...
+      i++;
+    }
+    voicing = util.rotateElements(voicing, inv * -1);
+  } else { // Move top notes down
+    while (i < voices - inv) {
+      voicing[voices - (i + 1)] = voicing[voices - (i + 1)].add(Teoria.interval('P-8'));
+      i++;
+    }
+    voicing = util.rotateElements(voicing, (voices - inv));
+  }
```
Comment (jmorganmartin) [inline src/App.js:223]: Calculates the new Invervals for the voicing array (by adding/subtracting 'P8' intvervals). It inverts in different directions for different inversions not for optimization, but to keep the voices as close to the root range as possible.  

Then the array is rotated so the notes are in the correct order. Looking at this now, it's probably possible to do both those steps within the while loops, but I didn't realize I'd also need to change the order until I ran into the "accidental bug", so the `rotateElements()` helper function/call was added later.

File: src/App.js [hunk 6]
```diff
@@ -158,10 +188,45 @@ function updateSettings(e) {
...
+  let simple_inversion = [];
+  voicing.forEach(interval => {
+    simple_inversion.push(interval.toString());
+  });
+
+  chord.voicing(simple_inversion);
+
+  return chord;
+}
+
```
Comment (jmorganmartin) [inline src/App.js]: Re-spells our inverted voicing as intervals. 
For example: 1st inversion chord goes from 
```
['P1', 'M3', 'M5']
``` 
to 
```
['M3', 'M5', 'P8']
```

File: src/App.js [hunk 7]
```diff
@@ -217,7 +290,7 @@ function getQuestion(i) {
     return (
       <div>
         On your own paper please spell out <strong>
-        {chord.toString().replace('M', '').replace('#', '♯').replace(/b/g, '♭')}</strong> in <strong>{clef}</strong> clef. 
+        {chord.toString().replace('M', '').replace('#', '♯').replace(/b/g, '♭')}, { (inversionType === 0) ? 'Root' : util.ordinalSuffix(inversionType) } inversion</strong> in <strong>{clef}</strong> clef.
```
Comment (jmorganmartin) [inline src/App.js:310]: I might suggest adding logic to hide "Root Inversion" if it's the only option selected (kinda redundant / potentially confusing). I didn't wanna get too detailed on that decision at this stage.

**Review by jmorganmartin — COMMENTED** [review]
I added some notes on the changes below

File: src/App.js [hunk 8]
```diff
@@ -158,10 +193,56 @@ function updateSettings(e) {
...
+  // Transpose chord up/down an octave if it breaks ledger line rules
+  // Note: certain triad inversions are impossible to place on certain staves without ledger lines:
+  // (e.g: in treble clef, all 1st-inversion A triads (and 2nd-inversion F triads)
+  // require a ledger line for either C4 or A5)
+  //
+  if (chord.notes()[chord.notes().length - 1].key(true) > Teoria.note(range[range.length - 1]).key(true) + 1) {
+    chord.transpose(Teoria.interval('P-8'));
+  } else if (chord.notes()[0].key(true) < Teoria.note(range[0]).key(true) - 1) {
+    chord.transpose(Teoria.interval('P8'));
+  }
```
Comment (jmorganmartin) [inline src/App.js:241]: This tries to make inversions follow ledger line rules (as much as possible).

**Review by stoyan — APPROVED** [review]

**Pull request merged**